middle_end/flambda2/cmx/flambda_cmx.ml

```diff
@@ -124,55 +190,90 @@ let compute_reachable_names_and_code ~module_symbol typing_env code =
+       ~used_closure_vars)
+
 let prepare_cmx_file_contents ~final_typing_env ~module_symbol
     ~used_closure_vars ~exported_offsets all_code =
   match final_typing_env with
   | None -> None
   | Some _ when Flambda_features.opaque () -> None
   | Some final_typing_env ->
-    let final_typing_env, canonicalise =
+    let typing_env, canonicalise =
       TE.Pre_serializable.create final_typing_env ~used_closure_vars
     in
-    let reachable_names =
-      compute_reachable_names_and_code ~module_symbol final_typing_env all_code
-    in
-    let all_code =
-      (* CR mshinwell: do we need to remove unused closure ID bindings from the
-         result types too? *)
-      all_code
-      |> EC.remove_unused_closure_vars_from_result_types_and_shortcut_aliases
-           ~used_closure_vars ~canonicalise
-      |> EC.remove_unreachable ~reachable_names
+    let create_typing_env reachable_names =
+      TE.Serializable.create typing_env ~reachable_names
     in
-    let final_typing_env =
-      TE.Serializable.create final_typing_env ~reachable_names
+    let free_names_of_name name =
+      Option.map T.free_names
+        (TE.Pre_serializable.find_or_missing typing_env name)
     in
-    (* We need to re-export offsets for everything reachable from the cmx file;
-       closure_vars/ids can be reachable from the typing env, and the exported
-       code.
+    prepare_cmx ~module_symbol create_typing_env ~free_names_of_name
+      ~used_closure_vars ~canonicalise ~exported_offsets all_code
 
-       In the case of the exported code, we already have offsets for all
-       closure_ids/vars reachable from the code of the current compilation unit,
-       but since we also re-export code from other compilation units, we need to
-       take those into account. *)
-    (* CR gbury: it might be more efficient to not compute the free names for
-       all exported code, but fold over the exported code to avoid allocating
-       some free_names *)
-    let free_names_of_all_code = EC.free_names all_code in
-    let closure_elts_used_in_typing_env =
-      TE.Serializable.free_closure_ids_and_closure_vars final_typing_env
+let prepare_cmx_from_approx ~approxs ~module_symbol ~exported_offsets
+    ~used_closure_vars all_code =
+  if Flambda_features.opaque ()
+  then None
+  else
+    let typing_env =
+      TE.Serializable.create_from_closure_conversion_approx approxs
     in
-    let exported_offsets =
-      exported_offsets
-      |> Closure_offsets.reexport_closure_ids
-           (Name_occurrences.closure_ids free_names_of_all_code)
-      |> Closure_offsets.reexport_closure_vars
-           (Name_occurrences.closure_vars free_names_of_all_code)
-      |> Closure_offsets.reexport_closure_ids
-           (Name_occurrences.closure_ids closure_elts_used_in_typing_env)
-      |> Closure_offsets.reexport_closure_vars
-           (Name_occurrences.closure_vars closure_elts_used_in_typing_env)
+    let create_typing_env _ = typing_env in
```

Comment:
We could actually use the `reachable_names` map to filter the approximations.
This is going to help in cases where symbols are used only during the toplevel module initialisation, and not exported in the signature.
```ocaml
let create_typing_env reachable_names =
  let approxs =
    Symbol.Map.filter (Name_occurrences.mem_symbol reachable_names) approxs
  in
  TE.Serializable.create_from_closure_conversion_approx approxs
in
```
In practice, this shouldn't make a huge difference as most symbols will be either a field of the global module or used in a reachable function, but it looks like a straightforward improvement made easier by this PR.